### PR TITLE
 Asynchronously produce the SpellChecker used for fuzzy matching.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
By doing this we don't block features that don't need the SpellChecker.
at the same time, because we're asynchronously producing it, once it is
ready, features that want it but don't want to block on it being produced
(like add-using) can light up.

I recommend reading this with ```?w=1``` as there are indentation changes that GitHub isn't good at diffing.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
retest vsi please

**Comment by CyrusNajmabadi** [comment]
retest vsi please

**Comment by CyrusNajmabadi** [comment]
retest vsi please

**Comment by CyrusNajmabadi** [comment]
tagging @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
@dotnet-bot retest prtest/win/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit32/3826/
// Retest reason:

**Comment by CyrusNajmabadi** [comment]
retest vsi please

File: src/Workspaces/Core/Portable/Utilities/BKTree.Serialization.cs [hunk 1]
```diff
@@ -29,27 +29,33 @@ internal void WriteTo(ObjectWriter writer)
...
+                    edges.Add(Edge.ReadFrom(reader));
+                }
+
+                return new BKTree(concatenatedLowerCaseWords, nodes.MoveToImmutable(), edges.MoveToImmutable());
             }
-
-            var edgeCount = reader.ReadInt32();
-            var edges = ImmutableArray.CreateBuilder<Edge>(edgeCount);
-            for (var i = 0; i < edgeCount; i++)
+            catch
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/Utilities/BKTree.Serialization.cs]: Ugh.  Can we at least use a filter to report a non-fatal watson here?

File: src/Workspaces/Core/Portable/Utilities/SpellChecker.cs [hunk 2]
```diff
@@ -34,12 +38,31 @@ public IList<string> FindSimilarWords(string value)
...
+                {
+                    var version = VersionStamp.ReadFrom(reader);
+                    var bkTree = BKTree.ReadFrom(reader);
+                    if (bkTree != null)
+                    {
+                        return new SpellChecker(version, bkTree);
+                    }
+                }
+            }
+            catch
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/Utilities/SpellChecker.cs]: And here?

File: src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Serialization.cs [hunk 3]
```diff
@@ -88,19 +116,22 @@ internal partial class SymbolTreeInfo : IObjectWritable
...
-                    return null;
+                    return default(T);
                 }
 
-                // compute it if we couldn't load it from cache
-                info = Create(version, assembly, cancellationToken);
-                if (info != null)
+                // Now, try to create a new instance and write it to the persistence service.
+                result = create(version);
+                if (result != null)
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Serialization.cs]: Above you switch from `null` to `default(T)`.  Why not here?

**Comment by Pilchie** [comment]
:-1: Really don't like the `catch { }`'s.

File: src/Workspaces/Core/Portable/Utilities/BKTree.Serialization.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Utilities/BKTree.Serialization.cs]: We don't in any other serialization code. I think the overall assumption is that we shouldn't assume data we cache on disk stays valid.

File: src/Features/Core/Portable/CodeFixes/AddImport/AbstractAddImportCodeFixProvider.cs [hunk 4]
```diff
@@ -119,12 +132,19 @@ internal abstract partial class AbstractAddImportCodeFixProvider<TSimpleNameSynt
...
             await FindResultsInAllProjectSymbolsAsync(project, allSymbolReferences, finder, exact).ConfigureAwait(false);
 
-            // Now search unreferenced projects, and see if they have any source symbols that match
-            // the search string.
-            await FindResultsInUnreferencedProjectSourceSymbolsAsync(projectToAssembly, project, allSymbolReferences, finder, exact, cancellationToken).ConfigureAwait(false);
+            // Only bother doing this for host workspaces.  We don't want this for 
+            // things like the Interactive workspace as we can't even add project
+            // references to the interactive window.  We could consider adding metadata
+            // references with #r in the future.
+            if (IsHostOrTestWorkspace(project))
```
Comment (dpoeschl) [inline src/Features/Core/Portable/CodeFixes/AddImport/AbstractAddImportCodeFixProvider.cs]: You already guarded the callsites of FindResults with the same check, right? Also, unrelated to your change, but FindResults --> FindResultsAsync.

File: src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs [hunk 5]
```diff
@@ -23,9 +23,16 @@ internal partial class SymbolTreeInfo
         private readonly IReadOnlyList<Node> _nodes;
 
         /// <summary>
-        /// The spell checker we use for fuzzy match queries.
+        /// The task that produces he spell checker we use for fuzzy match queries.
```
Comment (dpoeschl) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs]: he -> the

File: src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs [hunk 6]
```diff
@@ -23,9 +23,16 @@ internal partial class SymbolTreeInfo
...
+        /// We use a task so that we can generate the symbol tree info without having
```
Comment (dpoeschl) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs]: see cref="SymbolTreeInfo"

File: src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs [hunk 7]
```diff
@@ -215,7 +231,7 @@ private int BinarySearch(string name)
             return -1;
         }
 
-        #region Construction
+#region Construction
```
Comment (dpoeschl) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs]: Why? #regions are usually indented.

**Comment by dpoeschl** [comment]
Please fix up the commits before committing.

File: src/Features/Core/Portable/CodeFixes/AddImport/AbstractAddImportCodeFixProvider.cs (hunk 4 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/CodeFixes/AddImport/AbstractAddImportCodeFixProvider.cs]: No, on line 83 there's an unguarded call to FindResults.

File: src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs (hunk 7 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs]: This is whatever the formatter did.  I don't touch these :)   I also don't generally like or use regions in the first place.

**Comment by CyrusNajmabadi** [comment]
retest vsi please

**Comment by CyrusNajmabadi** [comment]
Please review again.  I've responded to the feedback about empty catches.  We now at least log telemetry for this so that we can react appropriately if we see a spike in exceptions thrown during deserialization.

**Comment by CyrusNajmabadi** [comment]
@Pilchie Let me know if this addressed your objections.

**Comment by CyrusNajmabadi** [comment]
@dotnet-bot retest prtest/win/vsi/p2 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn_prtest_win_vsi_p2/304/
// Retest reason: Known issue: https://github.com/dotnet/roslyn/issues/7441

**Comment by Pilchie** [comment]
Okay :+1: now.

**Pull request merged**